## Data analysis and manipulations using Pandas

**Agenda**

* Pandas Introduction
* Loading DataFrames from file data
* Preliminary data analysis
* Data type conversion
* Merging dataframes
* Handling duplicate data
* Handling missing data
* Identifying anamolies in the data
* Data aggregations
* Date and Time related data manipulations
* Understanding multi-indexing
* Pivot tables - data reshape
* Writing custom functions

### <font color='blue'>Problem Statement:</font>
A retail giant which operates a chain of hyper markets across the multiple cities wants to develop new strategies for improving the business. 

In this process, the marketing team has provided a data to data analytics team to understand the patterns in their customer transactions in addition to profiling the customers based on their demographics and transactions. 

The data analytics team is tasked to extract insights in the data which should help in developing new strategies for improving the business. 

High level overview of the data:
- Data has been provided for July- Sep 2018 Quarter which is from multiple data sources
  - `Demographics.csv`: Consists of demographic data of about 5.9k records that has  information about the customer demographics.
  - `Transactions.xlsx`: Consists of transactions data of about 260k records that has information about what product the customer purchased for what price
  - `Products.tsv`: Consists of products data of about 3.5k records which has a mapping of the product ids with product category.

- Given this data, we need to provide insights for developing new strategies
- We use basic statistics and aggregations in this context.

Datasets shared in compressed file or
In the shared server, data sets are available at the path **`/home/datasets/lab/`**.

Import libraries under the respective aliases

In [1]:
import os
import numpy as np
import pandas as pd

In [2]:
# data_path = "./datasets/"
# Comment above line and uncomment below line if you are working on shared server.
data_path = "/home/datasets/lab/"

In [3]:
os.listdir(data_path)

['pd_Transactions.xlsx',
 'New_Data.csv',
 'Housing_Data.csv',
 'CSE7215o_HOT_Data',
 'Transactions_New.csv',
 'CustomerData.csv',
 'Transactions.csv',
 'MEM_GermanData.csv',
 'pd_Products.tsv',
 'stuff.csv',
 'CombineLevels.csv',
 'CUTe',
 'GDPandSP500.csv',
 'Products.tsv',
 'German_Data.csv',
 'Data.rds',
 'Flights_Data.csv',
 'MergeData.xlsx',
 'pd_Demographics.csv',
 'cars.csv',
 'MEM_HeartData.csv',
 'pd_matches2008_2016.csv',
 'housingdata.csv',
 'Demographics.csv',
 'Ames_Housing_Data.csv',
 'Customer_Demographics_MV_DOB.csv',
 'data_SR.csv',
 'Transactions.xlsx',
 'MEM_HeartTestData.csv',
 'chickegg.csv',
 'Products.xlsx',
 'Customer_Bank Details_MV.csv',
 'pd_MergeData.xlsx',
 'RealEstateData.csv',
 'Tennis_Data.csv',
 'ReduceLevels.csv',
 'Salesdata_hypermart42.csv',
 'productWeeklyData.csv',
 'Flights_Delay.csv',
 'Eval.csv',
 'AmesHousingData.csv',
 'Data.csv',
 'Demographics.xlsx',
 'Bank.csv',
 'Eval2.csv',
 'Bank_Data.csv',
 'MEM_GermanTestData.csv',
 'pd_matches2017.cs

In [4]:
demo_path = os.path.join(data_path, "pd_Demographics.csv")
prod_path = os.path.join(data_path, "pd_Products.tsv")
txns_path = os.path.join(data_path, "pd_Transactions.xlsx")
merg_path = os.path.join(data_path, "pd_MergeData.xlsx")

In [5]:
demo_path

'/home/datasets/lab/pd_Demographics.csv'

In [6]:
!head /home/datasets/lab/pd_Demographics.csv

User_ID,Gender,Age,Occupation,City_Category,Stay_In_Current_City_Years,Marital_Status
1000001,F,15.0,10,A,2,0
1000002,M,69.0,16,C,4+,0
1000003,M,34.0,15,A,3,0
1000004,M,46.0,7,B,2,1
1000005,M,27.0,20,A,1,1
1000006,F,51.0,9,A,1,0
1000007,M,37.0,1,B,1,1
1000008,M,34.0,12,C,4+,1
1000009,M,30.0,17,C,0,0


In [7]:
!head /home/datasets/lab/pd_Products.tsv

Product_ID	Product_Category
P00000142	3
P00000242	2
P00000342	4
P00000442	4
P00000542	4
P00000642	1
P00000742	4
P00000842	2
P00000942	1


### Load data and create dataframes

- Read csv files using the read_csv function.
- Read tsv files using the read_csv function with a `tab \t` seperator.
- Read excel files using the read_excel function.

In [8]:
demographics = pd.read_csv(demo_path)
#demographics = pd.read_csv("/home/datasets/lab/Demographics.csv")

In [9]:
products = pd.read_csv(prod_path, sep = "\t")

In [10]:
transactions = pd.read_excel(txns_path, sheet_name="TransactionsData")

Display the sample content from each dataframe

In [11]:
demographics.head()

User_ID Gender   Age  Occupation City_Category Stay_In_Current_City_Years  \
0  1000001      F  15.0          10             A                          2   
1  1000002      M  69.0          16             C                         4+   
2  1000003      M  34.0          15             A                          3   
3  1000004      M  46.0           7             B                          2   
4  1000005      M  27.0          20             A                          1   

   Marital_Status  
0               0  
1               0  
2               0  
3               1  
4               1

In [12]:
products.head()

Product_ID  Product_Category
0  P00000142                 3
1  P00000242                 2
2  P00000342                 4
3  P00000442                 4
4  P00000542                 4

In [13]:
transactions.head()

Date  User_ID Product_ID  Purchase
0  28-9-2018  1000001  P00069042       837
1  29-7-2018  1000001  P00117942       884
2  28-7-2018  1000001  P00258742       691
3  27-7-2018  1000001  P00142242       788
4  29-7-2018  1000001  P00297042       784

List column names from each dataframe

In [14]:
demographics.columns

Index(['User_ID', 'Gender', 'Age', 'Occupation', 'City_Category',
       'Stay_In_Current_City_Years', 'Marital_Status'],
      dtype='object')

In [15]:
products.columns

Index(['Product_ID', 'Product_Category'], dtype='object')

In [16]:

transactions.columns

Index(['Date', 'User_ID', 'Product_ID', 'Purchase'], dtype='object')

### Preliminary data analysis

As the files are read, Do preliminary data analysis

Generally this consists of understanding the data.
- How many records and how many attributes are there?
- What are the datatypes of these attributes?
- Are there any missing values in the data?

To check the first and last 5 records of the data, we use head and tail
- dataframe.head() and 
- dataframe.tail(). 

We can specify a number in the parenthesis and those many records would be displayed.


In [17]:
demographics.head()

User_ID Gender   Age  Occupation City_Category Stay_In_Current_City_Years  \
0  1000001      F  15.0          10             A                          2   
1  1000002      M  69.0          16             C                         4+   
2  1000003      M  34.0          15             A                          3   
3  1000004      M  46.0           7             B                          2   
4  1000005      M  27.0          20             A                          1   

   Marital_Status  
0               0  
1               0  
2               0  
3               1  
4               1

In [18]:
demographics.tail()

User_ID Gender   Age  Occupation City_Category  \
5903  1007081      F   NaN          17             B   
5904  1007091      M  40.0           8             C   
5905  1007092      M  45.0           8             C   
5906  1007095      M  41.0           8             C   
5907  1007099      F  47.0           8             C   

     Stay_In_Current_City_Years  Marital_Status  
5903                         4+               0  
5904                         4+               1  
5905                         4+               1  
5906                         4+               1  
5907                         4+               1

Verify first 3 records

In [19]:
demographics.head(3)

User_ID Gender   Age  Occupation City_Category Stay_In_Current_City_Years  \
0  1000001      F  15.0          10             A                          2   
1  1000002      M  69.0          16             C                         4+   
2  1000003      M  34.0          15             A                          3   

   Marital_Status  
0               0  
1               0  
2               0

Verify the last 3 records

In [20]:
demographics.tail(3)

User_ID Gender   Age  Occupation City_Category  \
5905  1007092      M  45.0           8             C   
5906  1007095      M  41.0           8             C   
5907  1007099      F  47.0           8             C   

     Stay_In_Current_City_Years  Marital_Status  
5905                         4+               1  
5906                         4+               1  
5907                         4+               1

Verify random sample data

In [21]:
demographics.sample(10)

User_ID Gender   Age  Occupation City_Category  \
3711  1003803      F  41.0           9             C   
3078  1003155      M  28.0           7             A   
928   1000952      M  72.0          16             C   
3784  1003877      M  49.0           1             C   
1079  1001111      F  53.0          16             C   
2074  1002127      M  25.0           4             C   
3787  1003880      M  28.0           7             C   
1827  1001873      M  36.0           7             B   
3810  1003903      M  33.0           0             C   
505   1000511      F  50.0           4             C   

     Stay_In_Current_City_Years  Marital_Status  
3711                          0               0  
3078                          1               0  
928                           0               1  
3784                          1               1  
1079                          1               1  
2074                          0               1  
3787                          2               0  
1827                          2               0  
3810                          1               0  
505                           1               1

To get an overall understanding of data at a high level
- we use data.describe() function

describe produces multiple summary statistics in one shot.

In [22]:
demographics.describe()

User_ID          Age   Occupation  Marital_Status
count  5.908000e+03  5905.000000  5908.000000     5908.000000
mean   1.003027e+06    35.960711     8.155890        0.420785
std    1.749989e+03    12.986736     6.318067        0.493727
min    1.000001e+06    15.000000     0.000000        0.000000
25%    1.001515e+06    26.000000     3.000000        0.000000
50%    1.003026e+06    34.000000     7.000000        0.000000
75%    1.004536e+06    45.000000    14.000000        1.000000
max    1.007099e+06   129.000000    20.000000        1.000000

Observe that we have only the partial information. This is because, describe by default gives the summary of only numeric attributes. If we need to get information about all the attributes then we need to pass additional parameter to describe function.

On non-numeric data, describe produces alternative summary statistics.

In [23]:
demographics.describe(include="all")

User_ID Gender          Age   Occupation City_Category  \
count   5.908000e+03   5908  5905.000000  5908.000000          5908   
unique           NaN      2          NaN          NaN             3   
top              NaN      M          NaN          NaN             C   
freq             NaN   4235          NaN          NaN          3149   
mean    1.003027e+06    NaN    35.960711     8.155890           NaN   
std     1.749989e+03    NaN    12.986736     6.318067           NaN   
min     1.000001e+06    NaN    15.000000     0.000000           NaN   
25%     1.001515e+06    NaN    26.000000     3.000000           NaN   
50%     1.003026e+06    NaN    34.000000     7.000000           NaN   
75%     1.004536e+06    NaN    45.000000    14.000000           NaN   
max     1.007099e+06    NaN   129.000000    20.000000           NaN   

       Stay_In_Current_City_Years  Marital_Status  
count                        5908     5908.000000  
unique                          5             NaN  
top                             1             NaN  
freq                         2087             NaN  
mean                          NaN        0.420785  
std                           NaN        0.493727  
min                           NaN        0.000000  
25%                           NaN        0.000000  
50%                           NaN        0.000000  
75%                           NaN        1.000000  
max                           NaN        1.000000

In [24]:
demographics.describe(include="all").T

count unique  top  freq         mean       std  \
User_ID                     5908    NaN  NaN   NaN  1.00303e+06   1749.99   
Gender                      5908      2    M  4235          NaN       NaN   
Age                         5905    NaN  NaN   NaN      35.9607   12.9867   
Occupation                  5908    NaN  NaN   NaN      8.15589   6.31807   
City_Category               5908      3    C  3149          NaN       NaN   
Stay_In_Current_City_Years  5908      5    1  2087          NaN       NaN   
Marital_Status              5908    NaN  NaN   NaN     0.420785  0.493727   

                              min          25%          50%          75%  \
User_ID                     1e+06  1.00151e+06  1.00303e+06  1.00454e+06   
Gender                        NaN          NaN          NaN          NaN   
Age                            15           26           34           45   
Occupation                      0            3            7           14   
City_Category                 NaN          NaN          NaN          NaN   
Stay_In_Current_City_Years    NaN          NaN          NaN          NaN   
Marital_Status                  0            0            0            1   

                                   max  
User_ID                     1.0071e+06  
Gender                             NaN  
Age                                129  
Occupation                          20  
City_Category                      NaN  
Stay_In_Current_City_Years         NaN  
Marital_Status                       1

**Task**

Apply `describe` on a single column (Series) - Apply on `Age` column
result into `summary_stats_age`

What is the mean value

In [25]:
## Write 1 line of code.
# YOUR CODE HERE
summary_stats_age = demographics['Age'].describe()

In [26]:
summary_stats_age

count    5905.000000
mean       35.960711
std        12.986736
min        15.000000
25%        26.000000
50%        34.000000
75%        45.000000
max       129.000000
Name: Age, dtype: float64

In [27]:
# Check
print(round(summary_stats_age['mean']))

36.0


To get the dimensions of the data. To obtain how many attributes and records are available in the data
- We use dataframe.shape

In [28]:
### Get the dimensions of the data
demographics.shape

(5908, 7)

**Task** 

How many records and attributes are there in transactions and products dataframes?
- transactions dimensions into `trans_dims`
- products dimensions into `prod_dims`

In [29]:
## Write 2 lines of code.
# YOUR CODE HERE
trans_dims = transactions.shape
prod_dims = products.shape

In [30]:
## Check
print(transactions.shape)
print(products.shape)

(261507, 4)
(3449, 2)


- Observe that, all the three files correspond to transactions, products and demographics data based on the information from each customer, products they purchased and amount they spent. However, these are in different files and we need to consolidate and bring them into one single file, where we have all the attributes as columns and each record corresponds to one customer- their demographic information, products purchased and amount spent

- You all might have been to super markets right. If we observe the bill generated, each product we purchased would be printed in separate line i.e., 
    - If a customer purchases 10 products, there would be 10 records for that customer in transactions data, but the demographics information would still be one record. How would we consolidate these two files? We will look into this in a moment.

To observe the data types of each attribute
- We use dataframe.dtypes
- The data types given are how python interpreted them. Sometimes, they may be interpreted incorrectly. 

In such cases, we may have to change them to appropriate data types
    - Data types we have Integer/Numeric, Object (Category), Boolean.

In [31]:
# Understanding the data types for each data
demographics.dtypes

User_ID                         int64
Gender                         object
Age                           float64
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
dtype: object

#### Know the columns, memory information

In [32]:
# memory usage for object type is not counted

demographics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5908 entries, 0 to 5907
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   User_ID                     5908 non-null   int64  
 1   Gender                      5908 non-null   object 
 2   Age                         5905 non-null   float64
 3   Occupation                  5908 non-null   int64  
 4   City_Category               5908 non-null   object 
 5   Stay_In_Current_City_Years  5908 non-null   object 
 6   Marital_Status              5908 non-null   int64  
dtypes: float64(1), int64(3), object(3)
memory usage: 323.2+ KB


In [33]:
# memory usage for object type is now counted

demographics.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5908 entries, 0 to 5907
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   User_ID                     5908 non-null   int64  
 1   Gender                      5908 non-null   object 
 2   Age                         5905 non-null   float64
 3   Occupation                  5908 non-null   int64  
 4   City_Category               5908 non-null   object 
 5   Stay_In_Current_City_Years  5908 non-null   object 
 6   Marital_Status              5908 non-null   int64  
dtypes: float64(1), int64(3), object(3)
memory usage: 1.3 MB


### How do we differentiate between numeric and categorical attributes?

- Understading the variable names and doing a simple test. Does applying a statistical function like mean/average makes sense on that variable. If it does, then it is numeric else it is categorical
- Looking at the data. If a particular variable has only a few values which are repeating, probably they are categorical.

In [34]:
# convert the datatypes to appropriate type
demographics.Gender = demographics.Gender.astype('category')
demographics.Occupation = demographics.Occupation.astype('category')
demographics.City_Category = demographics.City_Category.astype('category')
demographics.Stay_In_Current_City_Years = demographics.Stay_In_Current_City_Years.astype('category')
demographics.Marital_Status = demographics.Marital_Status.astype('category')

In [35]:
demographics.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5908 entries, 0 to 5907
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype   
---  ------                      --------------  -----   
 0   User_ID                     5908 non-null   int64   
 1   Gender                      5908 non-null   category
 2   Age                         5905 non-null   float64 
 3   Occupation                  5908 non-null   category
 4   City_Category               5908 non-null   category
 5   Stay_In_Current_City_Years  5908 non-null   category
 6   Marital_Status              5908 non-null   category
dtypes: category(5), float64(1), int64(1)
memory usage: 123.1 KB


Total memory usage is reduced because the object type occupies more memory than category.

**Question: What are data types of products and transactions dataframes?**

In [36]:
print("TRANSACTIONS DATA TYPES")
print(transactions.dtypes)

TRANSACTIONS DATA TYPES
Date          object
User_ID        int64
Product_ID    object
Purchase       int64
dtype: object


In [37]:
print("PRODUCTS DATA TYPES")
print(products.dtypes)

PRODUCTS DATA TYPES
Product_ID          object
Product_Category     int64
dtype: object


- Observe that the Product Category data type is interpreted as integer but we know that it should be a category. Just because it was represented in numbers, it was interpreted as numeric. So we need to convert it to category

    - We use column.astype() function to change the data type

In [38]:
# We need to assign the change to the attribute and hence, syntax is as shown below
products['Product_Category'] = products['Product_Category'].astype('category')

# Verify the datatypes after changes
print(products.dtypes)

Product_ID            object
Product_Category    category
dtype: object


### Understand what is there in each data file

- Are there any duplicates in the data?

In [39]:
# Solution: To answer this question we need to see how many unique customers are 
# there and are these equal to total number of records

# Which does not match the total number of records (5908) in the data. 
# That means the User_ID are not unique and there duplicates.
demographics.User_ID.nunique() 

5900

- Observe that there are 5908 records in the customer data however there are only 5900 unique customers. What does this mean?
    - Are there any duplicates in the data. Is it ok to have duplicate records for the customers in this data?

In [40]:
# Analysing the duplicate records
duplicate_rows = demographics.duplicated(keep='first')
demographics[duplicate_rows]

User_ID Gender   Age Occupation City_Category  \
12    1000012      M  30.0         12             C   
66    1000067      F  55.0          5             C   
67    1000067      F  55.0          5             C   
68    1000067      F  55.0          5             C   
221   1000220      M  34.0         12             A   
222   1000220      M  34.0         12             A   
1126  1001158      F  39.0          1             C   
2151  1002203      M  49.0         12             B   

     Stay_In_Current_City_Years Marital_Status  
12                            2              0  
66                           4+              1  
67                           4+              1  
68                           4+              1  
221                           3              0  
222                           3              0  
1126                          2              1  
2151                         4+              1

In [41]:
# Ignoring the duplicate records
demo_no_duplicates = demographics.drop_duplicates(keep='first')
# demo_no_duplicates = demographics.[~demographics.duplicated()]

print(demo_no_duplicates.shape)
print(demo_no_duplicates['User_ID'].nunique())

(5900, 7)
5900


In [42]:
# Of these 5900 customers, how many are male and how many are female
demo_no_duplicates.Gender.value_counts()

M    4231
F    1669
Name: Gender, dtype: int64

In [43]:
demo_no_duplicates.Gender.value_counts(normalize=True)

M    0.717119
F    0.282881
Name: Gender, dtype: float64

- Observe that there are 4231 (≈72%) Males and 1669 (≈28%) Females. Males are nearly 2.5 times more than Females.

### Explore demographics data
Is customer data dominating by any particular demographic segment? Or uniform?

In [44]:
# What is the proportion of Male and Female?
demo_no_duplicates.Gender.value_counts(normalize=True) ## So we have about 72% male and 28% female.

M    0.717119
F    0.282881
Name: Gender, dtype: float64

Subset people whose Age is less than 30 years

In [45]:
demo_no_duplicates.Age < 30

0        True
1       False
2       False
3       False
4        True
        ...  
5903    False
5904    False
5905    False
5906    False
5907    False
Name: Age, Length: 5900, dtype: bool

In [46]:
subset_age_less_than_30 = demo_no_duplicates[demo_no_duplicates.Age < 30]
subset_age_less_than_30.head()

User_ID Gender   Age Occupation City_Category Stay_In_Current_City_Years  \
0   1000001      F  15.0         10             A                          2   
4   1000005      M  27.0         20             A                          1   
18  1000018      F  23.0          3             B                          3   
19  1000019      M  15.0         10             A                          3   
20  1000020      M  29.0         14             A                          0   

   Marital_Status  
0               0  
4               1  
18              0  
19              0  
20              0

In [47]:
subset_age_less_than_30.shape

(2125, 7)

**Question: How many customers are there by each City_Category?**

***These questions will allow us to analyse if there are more number of customers belonging to one particular city or it is uniform and so on***

In [48]:
demo_no_duplicates.City_Category.value_counts()

C    3144
B    1708
A    1048
Name: City_Category, dtype: int64

In [49]:
demo_no_duplicates.head()

User_ID Gender   Age Occupation City_Category Stay_In_Current_City_Years  \
0  1000001      F  15.0         10             A                          2   
1  1000002      M  69.0         16             C                         4+   
2  1000003      M  34.0         15             A                          3   
3  1000004      M  46.0          7             B                          2   
4  1000005      M  27.0         20             A                          1   

  Marital_Status  
0              0  
1              0  
2              0  
3              1  
4              1

In [50]:
demo_no_duplicates.groupby(['City_Category']).count()

User_ID  Gender   Age  Occupation  Stay_In_Current_City_Years  \
City_Category                                                                  
A                 1048    1048  1046        1048                        1048   
B                 1708    1708  1707        1708                        1708   
C                 3144    3144  3144        3144                        3144   

               Marital_Status  
City_Category                  
A                        1048  
B                        1708  
C                        3144

In [51]:
demo_no_duplicates.groupby(['City_Category'])['User_ID'].count()

City_Category
A    1048
B    1708
C    3144
Name: User_ID, dtype: int64

**Data Aggregations**
Aggregations refer to any data transformation that produces scalar values from
arrays. The preceding examples have used several of them, including mean, count,
min, and sum. Many common aggregations, 
have optimized implementations. 

- `count`: Number of non-NA values in the group
- `sum`: Sum of non-NA values
- `mean`: Mean of non-NA values
- `median`: Arithmetic median of non-NA values
- `std`, `var`: Unbiased (n – 1 denominator) standard deviation and variance
- `min`, `max`: Minimum and maximum of non-NA values
- `prod`: Product of non-NA values
- `first`, `last`: First and last non-NA values

**Task**

Which `City_Category` has the most number of people with `Occupation` value `4` 

result into `city_occupation_max_counts`

In [52]:
# YOUR CODE HERE
city_occupation_counts = demo_no_duplicates[demo_no_duplicates.Occupation == 4].\
    groupby('City_Category')['Age'].\
    count().\
    sort_values(ascending=False)
city_occupation_counts

City_Category
C    332
B    236
A    172
Name: Age, dtype: int64

In [53]:
city_occupation_max_counts =city_occupation_counts.reset_index()['City_Category'][0]

In [54]:
# Check
print(city_occupation_max_counts)

C


### Missing Value and Extreme Value analysis

- Whenever we get the data, we need to check if the data has any missing or extreme values/ anomalies in the data which might affect our analysis

In [55]:
demo_no_duplicates.isnull()

User_ID  Gender    Age  Occupation  City_Category  \
0       False   False  False       False          False   
1       False   False  False       False          False   
2       False   False  False       False          False   
3       False   False  False       False          False   
4       False   False  False       False          False   
...       ...     ...    ...         ...            ...   
5903    False   False   True       False          False   
5904    False   False  False       False          False   
5905    False   False  False       False          False   
5906    False   False  False       False          False   
5907    False   False  False       False          False   

      Stay_In_Current_City_Years  Marital_Status  
0                          False           False  
1                          False           False  
2                          False           False  
3                          False           False  
4                          False           False  
...                          ...             ...  
5903                       False           False  
5904                       False           False  
5905                       False           False  
5906                       False           False  
5907                       False           False  

[5900 rows x 7 columns]

In [56]:
demo_no_duplicates.isnull().sum()

User_ID                       0
Gender                        0
Age                           3
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64

In [57]:
demo_no_duplicates.isna().sum()

User_ID                       0
Gender                        0
Age                           3
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64

Observe that there are 3 missing values in the Age column. 

How to deal with these missing values?
  -  We ask the business if the data for these missing records available. If yes, then we can get this information from them else we need to look at other approaches.
  - One approach is to see what proportion of the data we have missing values. Is this a large value or a small value. If this value is very small compared to the number of records, we may choose to ignore these records. 
  -Other approach is to impute these missing values. 

**Dropping the records**

In [58]:
# Method 1
data_ig = demo_no_duplicates.dropna()
print("Records after removing rows with missing values are: ",data_ig.shape)
print("Records in original dataframe are: ",demo_no_duplicates.shape)

Records after removing rows with missing values are:  (5897, 7)
Records in original dataframe are:  (5900, 7)


In [59]:
print("Missing values in the dataframe after removal:")
print(data_ig.isna().sum())
print("\nMissing values in the original dataframe:")
print(demo_no_duplicates.isna().sum())

Missing values in the dataframe after removal:
User_ID                       0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64

Missing values in the original dataframe:
User_ID                       0
Gender                        0
Age                           3
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64


In [60]:
data_ig["Age"].describe()

count    5897.000000
mean       35.949975
std        12.986988
min        15.000000
25%        26.000000
50%        33.000000
75%        45.000000
max       129.000000
Name: Age, dtype: float64

**Imputing the missing values**
  - Imputing is basically a method of "guessing" the value in place of missing value. One method of imputation is central imputation where you replace the null value with the mean value of the column
  - But before imputing, we need to check if there are any extreme values in the data because the extreme values can impact the mean values.

Observe that there are some points above 100 and the max is about 129 (recall that we got the max value when we used the describe function on the data. These values could be anamolies. One of the approaches to tackle these extremevalues isto ignore the records before imputing the data. We can discuss other methods later.

In [61]:
# Removing the extreme values
# demo_no_duplicates[demo_no_duplicates.Age <= 100].shape
# demo_no_duplicates[~(demo_no_duplicates.Age > 100)].shape
data1 = demo_no_duplicates[~demo_no_duplicates.Age.isin(range(100,200))]
print("Dimensions of the processed data: ", data1.shape)
print("Missing values in Data processed", data1.isna().sum())

Dimensions of the processed data:  (5898, 7)
Missing values in Data processed User_ID                       0
Gender                        0
Age                           3
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64


In [62]:
# Lets describe the data again
print(data1['Age'].describe())

count    5895.000000
mean       35.922477
std        12.901177
min        15.000000
25%        26.000000
50%        33.000000
75%        45.000000
max        75.000000
Name: Age, dtype: float64


In [63]:
# Now let's impute the age
print("The number of missing values before imputation: ",data1.isna().sum())
print("\n")
data1['Age'].fillna(data1['Age'].mean(), inplace=True)

print(data1.Age.describe())
print("\n")
print("The number of missing values after imputation: ",data1.isna().sum())

The number of missing values before imputation:  User_ID                       0
Gender                        0
Age                           3
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64


count    5898.000000
mean       35.922477
std        12.897895
min        15.000000
25%        26.000000
50%        33.000000
75%        45.000000
max        75.000000
Name: Age, dtype: float64


The number of missing values after imputation:  User_ID                       0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
dtype: int64


/usr/share/anaconda3/lib/python3.7/site-packages/pandas/core/series.py:4523: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [64]:
# Fill with a value
# data1.fillna(0)
# Fill with different values for different columns
# data1.fillna({'Age': 0, 'Gender': 'M'})
# Fill using ffill
# data1.fillna(method='ffill')

In [65]:
# data1.fillna() # shift + tab for the help

 ### Discretization and Binning

Continuous data is often discretized or otherwise separated into “bins” for analysis.

Suppose if we want to group Age into discrete age buckets.

- Age Group binning
    - Age is numeric data. But the question we need to ask is does a 50 year old person behaviur would be different from 51 year old. Or a 21 year old behaviour be different from 22 year old. If not, then we need not want individual information on age but we need a collective information like what products are being purchased by a specific age group ( and not individual)
    - In order to group the customers based on Age we create our custom bins like 0-17 years, 18-25 years, 26-35 years, 36-45 years, 46-50 years, 51-55 years and 55+.

In [66]:
bins = [0, 17, 25, 35, 45, 50, 55, 100]
labels=['0-17','18-25','26-35','36-45','46-50','51-55','55+'] 

In [67]:
data1['Age_Group'] = pd.cut(data1['Age'], bins=bins, labels=labels)

/usr/share/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [68]:
data1.head()

User_ID Gender   Age Occupation City_Category Stay_In_Current_City_Years  \
0  1000001      F  15.0         10             A                          2   
1  1000002      M  69.0         16             C                         4+   
2  1000003      M  34.0         15             A                          3   
3  1000004      M  46.0          7             B                          2   
4  1000005      M  27.0         20             A                          1   

  Marital_Status Age_Group  
0              0      0-17  
1              0       55+  
2              0     26-35  
3              1     46-50  
4              1     26-35

**Question: Is there are particular age group that are more in number or is it uniform?**

In [69]:
data1.groupby(['Age_Group'])['User_ID'].count()

Age_Group
0-17      218
18-25    1069
26-35    2053
36-45    1173
46-50     532
51-55     494
55+       359
Name: User_ID, dtype: int64

In [70]:
data1['Age_Group'].value_counts()

26-35    2053
36-45    1173
18-25    1069
46-50     532
51-55     494
55+       359
0-17      218
Name: Age_Group, dtype: int64

In [71]:
data1['Age_Group'].value_counts(normalize = True)

26-35    0.348084
36-45    0.198881
18-25    0.181248
46-50    0.090200
51-55    0.083757
55+      0.060868
0-17     0.036962
Name: Age_Group, dtype: float64

Observe that people in the age group of 26-35 followed by 36-45 are more in number compared to other groups.


**Pair-wise frequencies**
- Let's visualize each of these in the form of a table. In excel we have a pivot table to do such analysis, here we have a crosstab
    - How many customers from each Age_Group are there in each City_Category.



In [72]:
pd.crosstab(data1['Age_Group'], data1['City_Category'])

City_Category    A    B    C
Age_Group                   
0-17            25   50  143
18-25          214  331  524
26-35          461  652  940
36-45          178  336  659
46-50           53  146  333
51-55           69  136  289
55+             47   57  255

Observe that City_Category C has more customers in all the age groups

**Question: How many customers of each gender belonging to each occupation?**

In [73]:
pd.crosstab(data1['Gender'], data1['Occupation'])

Occupation    0    1    2   3    4   5    6    7   8   9  ...   11   12   13  \
Gender                                                    ...                  
F           226  203   88  98  229  31   99  137   4  85  ...   22   46   33   
M           462  314  168  72  512  80  129  532  17   3  ...  106  330  107   

Occupation   14   15   16   17  18  19   20  
Gender                                       
F            78   28   49   51   4  15   77  
M           216  113  186  441  63  56  196  

[2 rows x 21 columns]

In [74]:
data1.groupby('Occupation')['Gender'].value_counts()

Occupation  Gender
0           M         462
            F         226
1           M         314
            F         203
2           M         168
            F          88
3           F          98
            M          72
4           M         512
            F         229
5           M          80
            F          31
6           M         129
            F          99
7           M         532
            F         137
8           M          17
            F           4
9           F          85
            M           3
10          M         126
            F          66
11          M         106
            F          22
12          M         330
            F          46
13          M         107
            F          33
14          M         216
            F          78
15          M         113
            F          28
16          M         186
            F          49
17          M         441
            F          51
18          M          63
            F      

In [75]:
pd.set_option('display.max_columns', None)

In [76]:
pd.crosstab(data1['Gender'], data1['Occupation'])

Occupation    0    1    2   3    4   5    6    7   8   9   10   11   12   13  \
Gender                                                                         
F           226  203   88  98  229  31   99  137   4  85   66   22   46   33   
M           462  314  168  72  512  80  129  532  17   3  126  106  330  107   

Occupation   14   15   16   17  18  19   20  
Gender                                       
F            78   28   49   51   4  15   77  
M           216  113  186  441  63  56  196

In [77]:
pd.crosstab(data1['Gender'], data1['Marital_Status'])

Marital_Status     0     1
Gender                    
F                948   721
M               2471  1758

**Task**

Find the `male`, `female` counts for each `City_Category` into `mf_counts_in_each_city`

In [78]:
## Write 1 line of code.
# YOUR CODE HERE
mf_counts_in_each_city = pd.crosstab(data1['Gender'], data1['City_Category'])

In [79]:
# Check
print(mf_counts_in_each_city)

City_Category    A     B     C
Gender                        
F              296   504   869
M              751  1204  2274


Observe that in both Male and Female, we have more number of unmarried customers than married.


### Transactions data analysis.

Find high revenue generating customers?

In [80]:
transactions.head()

Date  User_ID Product_ID  Purchase
0  28-9-2018  1000001  P00069042       837
1  29-7-2018  1000001  P00117942       884
2  28-7-2018  1000001  P00258742       691
3  27-7-2018  1000001  P00142242       788
4  29-7-2018  1000001  P00297042       784

In [81]:
high_revenue_customers = transactions.groupby(['User_ID'])['Purchase'].sum()

In [82]:
high_revenue_customers.head()

User_ID
1000001     9979
1000002    28416
1000003     7329
1000005    38641
1000006    14162
Name: Purchase, dtype: int64

In [83]:
# Top 10 revenue generating customers
high_revenue_customers.sort_values(ascending = False).head(10)

User_ID
1001680    401302
1004277    382836
1001015    317057
1001980    284272
1003224    282424
1005831    281230
1003391    281033
1001181    280374
1003618    274230
1000889    269989
Name: Purchase, dtype: int64

**Question: High revenue products?**

In [84]:
# Sorting the products based on descending order of purchase
high_revenue_products = transactions.groupby(['Product_ID'])['Purchase'].sum()
high_revenue_products.sort_values(ascending=False)[:10]

Product_ID
P00265242    1364054
P00058042    1037815
P00117942     918574
P00051442     916621
P00031042     889063
P00220442     878291
P00278642     853454
P00034742     828964
P00117442     825619
P00251242     777853
Name: Purchase, dtype: int64

Sort the transactions data based on Date and UserID

In [85]:
# sorting by Date and UserID
transactions.sort_values(['Date','User_ID'])

Date  User_ID Product_ID  Purchase
18896   1-7-2018  1000466  P00007142       686
18961   1-7-2018  1000470  P00003942       718
18990   1-7-2018  1000471  P00109742       978
19030   1-7-2018  1000474  P00090842       760
19199   1-7-2018  1000477  P00143342       772
...          ...      ...        ...       ...
261116  9-9-2018  1006036  P00050242       808
261151  9-9-2018  1006036  P00173842       804
261215  9-9-2018  1006036  P00089142       695
261453  9-9-2018  1006040  P00244042       829
261498  9-9-2018  1007091  P00123142       597

[261507 rows x 4 columns]

**Question: Find top 3 revenue generating customers?**

In [178]:
highest_revenue_customer = transactions.\
            groupby('User_ID', as_index=False).\
            agg({'Purchase': 'sum'})
highest_revenue_customer

Purchase  Product_ID
User_ID                      
1000001      9979          12
1000002     28416          37
1000003      7329          10
1000005     38641          55
1000006     14162          19
...           ...         ...
1007081      3550           1
1007091      5701           4
1007092      6050           2
1007095      1750           1
1007099      4847           3

[5885 rows x 2 columns]

In [87]:
highest_revenue_customer.sort_values('Purchase', ascending=False).head(3)

User_ID  Purchase
1631  1001680    401302
4157  1004277    382836
979   1001015    317057

**Question: Find the highest revenue generating customer for the product line **`P00143342`****?

In [88]:
# P00143342_data = transactions[transactions.Product_ID == 'P00143342']
P00143342_data = transactions[transactions['Product_ID'] == 'P00143342']

P00143342_data.groupby('User_ID', as_index=False).\
    agg({'Purchase': 'sum'}).\
    sort_values('Purchase').\
    tail(1)

User_ID  Purchase
48  1003778       834

In [89]:
# another way to acheive the same

P00143342_data = transactions.\
    groupby(['User_ID','Product_ID'], as_index=False).\
    agg({'Purchase':'sum'})

P00143342_data[P00143342_data.Product_ID == 'P00143342'].\
    sort_values('Purchase').\
    tail(1)

User_ID Product_ID  Purchase
166527  1003778  P00143342       834

**Question: Find the highest purchase details such as who made the purchase, on which product and when?**

In [90]:
highest_details = transactions.\
        groupby('User_ID').\
        max()

In [91]:
highest_details.head()

Date Product_ID  Purchase
User_ID                                
1000001  30-9-2018  P00297042       995
1000002  30-9-2018  P00356742       996
1000003  30-8-2018  P00343042       889
1000005  30-9-2018  P00355142      1000
1000006  30-9-2018   P0097942       999

In [92]:
highest_details.\
    reset_index().\
    sort_values('Purchase', ascending=False).\
    head(1)

User_ID      Date Product_ID  Purchase
5879  1007028  8-9-2018   P1005887      3850

In [93]:
transactions.shape

(261507, 4)

In [94]:
products.shape

(3449, 2)

In [95]:
transactions.Product_ID.nunique()

3264

In [96]:
products.Product_ID.nunique()

3449

Not all the products are involved in transactions this may be because some products are not sold in the given time period.

In [97]:
products.Product_ID.unique()

array(['P00000142', 'P00000242', 'P00000342', ..., 'P1003221', 'P1003381',
       'P1005887'], dtype=object)

In [98]:
# Which products have zero transactions

notransprod_list = list(set(products.Product_ID.unique()) - set(transactions.Product_ID.unique()))

In [99]:
notransprod_list

['P00298342',
 'P00126742',
 'P00296842',
 'P00188642',
 'P00321242',
 'P00013442',
 'P00213842',
 'P00073142',
 'P00159642',
 'P00363542',
 'P00185942',
 'P00149742',
 'P00039142',
 'P00053242',
 'P00238242',
 'P00152142',
 'P00325342',
 'P00172242',
 'P0093742',
 'P00348142',
 'P00369642',
 'P00364542',
 'P00290642',
 'P00353042',
 'P00202842',
 'P00012942',
 'P00307842',
 'P00299042',
 'P00360542',
 'P00341142',
 'P00201942',
 'P00065842',
 'P00166442',
 'P00357542',
 'P00274742',
 'P00106342',
 'P00107842',
 'P00310942',
 'P00194942',
 'P00077242',
 'P00353742',
 'P00272342',
 'P00090242',
 'P00150242',
 'P00262842',
 'P00013342',
 'P00308042',
 'P00054642',
 'P00074542',
 'P00011242',
 'P00107942',
 'P00091542',
 'P00301442',
 'P00106842',
 'P00295642',
 'P00063442',
 'P00245742',
 'P00352042',
 'P00107342',
 'P00342442',
 'P00312042',
 'P00162842',
 'P00200542',
 'P00062342',
 'P00337042',
 'P00306742',
 'P00104042',
 'P00329142',
 'P00172642',
 'P00252342',
 'P00135942',
 'P0020

**Question: Which Product Category has more number of products with zero transactions**

In [100]:
notransprod_df = products[products.Product_ID.isin(notransprod_list)]
notransprod_df

Product_ID Product_Category
106   P00011242                6
117   P00012342                7
118   P00012442                6
122   P00012942                4
124   P00013242                6
...         ...              ...
3348  P00365942                7
3376  P00368742               10
3385  P00369642               10
3396   P0093742                5
3421   P0097542                6

[185 rows x 2 columns]

In [101]:
# add total number of products as column, 
prod_cats = notransprod_df.groupby('Product_Category', as_index=False).count()

prod_cats = prod_cats.rename({
    'Product_ID':'zero_transaction_products'
},
axis=1)

prod_cats.sort_values('zero_transaction_products', ascending = False)

Product_Category  zero_transaction_products
5                 6                         41
6                 7                         35
9                10                         23
3                 4                         19
0                 1                         18
4                 5                         14
11               12                         14
8                 9                          8
12               13                          6
1                 2                          4
2                 3                          1
7                 8                          1
13               14                          1
10               11                          0

In [102]:
# product categories with count of products
tmp2 = products.groupby('Product_Category', as_index=False).count()
tmp2 = tmp2.rename({
    'Product_ID':'total_products',
}, axis=1)

tmp2

Product_Category  total_products
0                 1             491
1                 2             152
2                 3              90
3                 4             966
4                 5             119
5                 6             102
6                 7            1047
7                 8               2
8                 9              25
9                10             254
10               11               4
11               12              88
12               13              98
13               14              11

Since there are so many products, instead of going with product_id, can we go with product category. For this, we may need to merge the data of POS with products so that we can map each product to their respective product category
- Let's understand the merge

### Example

Read merge dataset

In [103]:
Age       = pd.read_excel(merg_path, sheet_name="Ages")
Education = pd.read_excel(merg_path, sheet_name="Education")

In [104]:
Age

Name  Age
0  Alpha   21
1   Beta   28
2  Gamma   25
3  Delta   29

In [105]:
Education

Name  Education
0  Alpha  Bachelors
1  Kappa  Doctorate
2    Eta    Masters
3  Delta  Bachelors

### Combining and Merging Datasets

Data contained in pandas objects can be combined together in a number of ways.
- `pandas.merge` connects rows in DataFrames based on one or more keys. This will be familiar to users of SQL or other relational databases, as it implements
database join operations.
- `pandas.concat` concatenates or “stacks” together objects along an axis.
- The `combine_first` instance method enables splicing together overlapping data to fill in missing values in one object with values from another.

**Database-Style DataFrame Joins**

Merge or join operations combine datasets by linking rows using one or more keys.
These operations are central to relational databases (e.g., SQL-based). The merge
function in pandas is the main entry point for using these algorithms on your data.

In [106]:
combined_inner=pd.merge(Age, Education, on='Name')
combined_inner

Name  Age  Education
0  Alpha   21  Bachelors
1  Delta   29  Bachelors

By default merge does an ``inner`` join.

The keys in the result are the intersection,
or the common set found in both tables. 

Other possible options are `left`, `right`, and `outer`. 

The `outer join` takes the union of the keys, combining the
effect of applying both left and right joins.

***Different join types with how argument***
- `inner` Use only the key combinations observed in both tables
- `left` Use all key combinations found in the left table
- `right` Use all key combinations found in the right table
- `outer` Use all key combinations observed in both tables together


In [107]:
combined_outer = pd.merge(Age, Education, on='Name', how='outer')
combined_outer

Name   Age  Education
0  Alpha  21.0  Bachelors
1   Beta  28.0        NaN
2  Gamma  25.0        NaN
3  Delta  29.0  Bachelors
4  Kappa   NaN  Doctorate
5    Eta   NaN    Masters

In [108]:
combined_left = pd.merge(Age, Education, on='Name', how='left')
combined_left

Name  Age  Education
0  Alpha   21  Bachelors
1   Beta   28        NaN
2  Gamma   25        NaN
3  Delta   29  Bachelors

In [109]:
combined_right = pd.merge(Age, Education, on='Name', how='right')
combined_right

Name   Age  Education
0  Alpha  21.0  Bachelors
1  Kappa   NaN  Doctorate
2    Eta   NaN    Masters
3  Delta  29.0  Bachelors

***merge function arguments***
- `left` DataFrame to be merged on the left side.
- `right` DataFrame to be merged on the right side.
- `how` One of `inner`, `outer`, `left`, or `right`; defaults to `inner`.
- `on` Column names to join on. Must be found in both DataFrame objects. If not specified and no other join keys
given, will use the intersection of the column names in left and right as the join keys.
- `left_on` Columns in left DataFrame to use as join keys.
- `right_on` Analogous to left_on for left DataFrame.
- `left_index` Use row index in left as its join key (or keys, if a MultiIndex).
- `right_index` Analogous to left_index.
- `sort` Sort merged data lexicographically by join keys; True by default (disable to get better performance in
some cases on large datasets).
- `suffixes` Tuple of string values to append to column names in case of overlap; defaults to ('_x', '_y') (e.g., if
'data' in both DataFrame objects, would appear as 'data_x' and 'data_y' in result).
- `copy` If False, avoid copying data into resulting data structure in some exceptional cases; by default always
copies.
- `indicator` Adds a special column _merge that indicates the source of each row; values will be

**Concatenating Along an Axis**

In [110]:
df1 = pd.DataFrame(np.arange(6).reshape(3, 2), 
                   index=['a', 'b', 'c'],
                   columns=['one', 'two'])

df1

one  two
a    0    1
b    2    3
c    4    5

In [111]:
df2 = pd.DataFrame(5 + np.arange(4).reshape(2, 2), 
                   index=['a', 'c'],
                   columns=['three', 'four'])

df2

three  four
a      5     6
c      7     8

In [112]:
# Concatenate along rows
pd.concat([df1, df2], axis=0, sort=False)

one  two  three  four
a  0.0  1.0    NaN   NaN
b  2.0  3.0    NaN   NaN
c  4.0  5.0    NaN   NaN
a  NaN  NaN    5.0   6.0
c  NaN  NaN    7.0   8.0

In [113]:
# Concatenate along columns
pd.concat([df1, df2], axis=1, sort=False)

one  two  three  four
a    0    1    5.0   6.0
b    2    3    NaN   NaN
c    4    5    7.0   8.0

**Merge transactions and products dataframes**

In [114]:
transactions.columns

Index(['Date', 'User_ID', 'Product_ID', 'Purchase'], dtype='object')

In [115]:
products.columns

Index(['Product_ID', 'Product_Category'], dtype='object')

In [116]:
# Merging the data frames
prod_trans = pd.merge(transactions, products, on='Product_ID', how='left')

In [117]:
prod_trans.head()

Date  User_ID Product_ID  Purchase Product_Category
0  28-9-2018  1000001  P00069042       837                3
1  29-7-2018  1000001  P00117942       884                4
2  28-7-2018  1000001  P00258742       691                4
3  27-7-2018  1000001  P00142242       788                7
4  29-7-2018  1000001  P00297042       784                7

In [118]:
prod_trans.dtypes

Date                  object
User_ID                int64
Product_ID            object
Purchase               int64
Product_Category    category
dtype: object

In [119]:
prod_trans['Product_Category'].value_counts()

4     123311
7      92941
1      16842
10      9681
2       8167
3       6400
5       1360
13      1069
12      1047
9        216
6        216
14       164
8         81
11        12
Name: Product_Category, dtype: int64

**Question: How many products are there in each product category?**

In [120]:
prod_trans.groupby(['Product_Category'])['Product_ID'].nunique().sort_values(ascending = False)

Product_Category
7     1012
4      947
1      473
10     231
2      148
5      105
13      92
3       89
12      74
6       61
9       17
14      10
11       4
8        1
Name: Product_ID, dtype: int64

Observe that more number of products are available in product category 7

**Question: High revenue product categories?**

In [121]:
high_revenue_prod_cats = prod_trans.groupby(['Product_Category'])['Purchase'].sum()
high_revenue_prod_cats.sort_values(ascending = False)[:10]

Product_Category
4     85083356
7     73176675
1     13403429
2      7036410
10     6322255
3      4851457
5      1136601
13      904098
12      892889
9       193190
Name: Purchase, dtype: int64

Observe that though there are more number of products in category 7, the highest revenue is from category 4.

**Task**

What is the `Product_Category` with the highest average `sale amount (Purchase)` into `prod_highest_avg_sale`?

In [122]:
## Write 1 or 2 lines of code.
# YOUR CODE HERE
avg_revenue_prod_cat = prod_trans.groupby('Product_Category')['Purchase'].mean()

In [123]:
prod_highest_avg_sale = avg_revenue_prod_cat.sort_values(ascending=False)[0]

In [124]:
# Check
print(prod_highest_avg_sale)

2887.5


### Understanding anomalies in the data

In [125]:
prod_trans.Z_score = (prod_trans.Purchase - prod_trans.Purchase.mean())/prod_trans.Purchase.std()

/usr/share/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  """Entry point for launching an IPython kernel.


In [126]:
prod_trans.head()

Date  User_ID Product_ID  Purchase Product_Category
0  28-9-2018  1000001  P00069042       837                3
1  29-7-2018  1000001  P00117942       884                4
2  28-7-2018  1000001  P00258742       691                4
3  27-7-2018  1000001  P00142242       788                7
4  29-7-2018  1000001  P00297042       784                7

In [127]:
prod_trans['Z_score'] = (prod_trans.Purchase - prod_trans.Purchase.mean())/prod_trans.Purchase.std()

In [128]:
prod_trans.head()

Date  User_ID Product_ID  Purchase Product_Category   Z_score
0  28-9-2018  1000001  P00069042       837                3  0.693665
1  29-7-2018  1000001  P00117942       884                4  1.028394
2  28-7-2018  1000001  P00258742       691                4 -0.346133
3  27-7-2018  1000001  P00142242       788                7  0.344692
4  29-7-2018  1000001  P00297042       784                7  0.316204

In [129]:
print("Dimensions: ")
print(prod_trans.shape)

# Find outliers +/- 3 
Outliers = prod_trans[(np.abs(prod_trans['Z_score']) > 3)].index
# Outliers count
print(len(Outliers))
Outliers

Dimensions: 
(261507, 6)
12


Int64Index([261488, 261489, 261490, 261491, 261492, 261493, 261496, 261497,
            261501, 261502, 261503, 261506],
           dtype='int64')

In [130]:
prod_trans.iloc[Outliers].head(10)

Date  User_ID Product_ID  Purchase Product_Category    Z_score
261488  16-8-2018  1007021   P1003112      2500               11  12.537388
261489   6-7-2018  1007021   P1005887      3850               11  22.151956
261490  13-8-2018  1007021   P1003381      1750               11   7.195962
261491   5-8-2018  1007028   P1003381      1750               11   7.195962
261492  11-7-2018  1007028   P1003112      2500               11  12.537388
261493  18-7-2018  1007028   P1005887      3850               11  22.151956
261496  12-9-2018  1007081   P1003221      3550               11  20.015386
261497  10-7-2018  1007091   P1003221      3550               11  20.015386
261501  20-9-2018  1007092   P1003221      3550               11  20.015386
261502  12-9-2018  1007092   P1003112      2500               11  12.537388

In [131]:
print(prod_trans['Z_score'].min())
print(prod_trans['Z_score'].max())

-1.692172360600414
22.151956419617328


- Observe that all the records that have very high value of purchase belong to item category 11
- With this observation, what can be done? Can we include category 11 with all other categories and perform the analysis or since they are very different (in terms of revenue) should they be analysed separately?

### Date and Time related data

In [132]:
prod_trans.head()

Date  User_ID Product_ID  Purchase Product_Category   Z_score
0  28-9-2018  1000001  P00069042       837                3  0.693665
1  29-7-2018  1000001  P00117942       884                4  1.028394
2  28-7-2018  1000001  P00258742       691                4 -0.346133
3  27-7-2018  1000001  P00142242       788                7  0.344692
4  29-7-2018  1000001  P00297042       784                7  0.316204

In [133]:
prod_trans.dtypes

Date                  object
User_ID                int64
Product_ID            object
Purchase               int64
Product_Category    category
Z_score              float64
dtype: object

Observe, there is a date column which is interpreted as an object. 

We need to convert this to date format for further manipulation.

In [134]:
# Coversion of date to date time object
prod_trans['Date'] = pd.to_datetime(prod_trans['Date'], format="%d-%m-%Y")
prod_trans.dtypes

Date                datetime64[ns]
User_ID                      int64
Product_ID                  object
Purchase                     int64
Product_Category          category
Z_score                    float64
dtype: object

In [135]:
# Extract the week and months from the data
prod_trans['Week'] = prod_trans['Date'].dt.week
prod_trans['Month'] = prod_trans['Date'].dt.month
prod_trans.head(5)

/usr/share/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  


Date  User_ID Product_ID  Purchase Product_Category   Z_score  Week  \
0 2018-09-28  1000001  P00069042       837                3  0.693665    39   
1 2018-07-29  1000001  P00117942       884                4  1.028394    30   
2 2018-07-28  1000001  P00258742       691                4 -0.346133    30   
3 2018-07-27  1000001  P00142242       788                7  0.344692    30   
4 2018-07-29  1000001  P00297042       784                7  0.316204    30   

   Month  
0      9  
1      7  
2      7  
3      7  
4      7

**Question: Which month have highest amount of sales? (sales value)**

In [136]:
sales_by_month_amt = prod_trans.groupby(['Month'])['Purchase'].sum()
sales_by_month_amt.sort_values(ascending = False)

Month
7    64918618
9    64283305
8    64208962
Name: Purchase, dtype: int64

**Question: Which month have highest amount of sales? (sales transactions)**

In [137]:
sales_by_month_volume = prod_trans.groupby(['Month'])['Purchase'].count()
sales_by_month_volume.sort_values(ascending = False)

Month
7    87791
9    86927
8    86789
Name: Purchase, dtype: int64

### Pivot

A pivot table is a table of statistics that summarizes the data of a more extensive table (such as from a database, spreadsheet, or business intelligence program). This summary might include sums, averages, or other statistics, which the pivot table groups together in a meaningful way.

Pivot tables are a technique in data processing. They arrange and rearrange (or "pivot") statistics in order to draw attention to useful information.

#### Pivot - Long to Wide format

In [138]:
# To identify if the person has a favorite product category or explore all categories
user_profile1 = prod_trans.groupby(['User_ID','Product_Category'])['Product_ID'].count()
user_profile1.head()

User_ID  Product_Category
1000001  1                   0
         2                   0
         3                   3
         4                   2
         5                   0
Name: Product_ID, dtype: int64

In [139]:
user_profile1 = user_profile1.reset_index()
user_profile1.head()

User_ID Product_Category  Product_ID
0  1000001                1           0
1  1000001                2           0
2  1000001                3           3
3  1000001                4           2
4  1000001                5           0

In [140]:
# Casting the data so that we know how many product categories user brought
user_pivot = user_profile1.pivot(index='User_ID', 
                                 columns='Product_Category',
                                 values="Product_ID")

# user_pivot = user_pivot.fillna(0)

In [141]:
user_pivot

Product_Category  1  2  3   4  5  6   7  8  9  10  11  12  13  14
User_ID                                                          
1000001           0  0  3   2  0  0   7  0  0   0   0   0   0   0
1000002           4  0  0  12  1  0  20  0  0   0   0   0   0   0
1000003           1  0  0   8  0  0   1  0  0   0   0   0   0   0
1000005           5  0  0  16  2  1  29  0  0   1   0   0   1   0
1000006           0  1  0  11  1  0   6  0  0   0   0   0   0   0
...              .. .. ..  .. .. ..  .. .. ..  ..  ..  ..  ..  ..
1007081           0  0  0   0  0  0   0  0  0   0   1   0   0   0
1007091           0  0  0   2  0  0   0  0  0   1   1   0   0   0
1007092           0  0  0   0  0  0   0  0  0   0   2   0   0   0
1007095           0  0  0   0  0  0   0  0  0   0   1   0   0   0
1007099           0  0  0   1  0  0   0  0  0   1   1   0   0   0

[5885 rows x 14 columns]

In [142]:
user_pivot.index

Int64Index([1000001, 1000002, 1000003, 1000005, 1000006, 1000007, 1000008,
            1000009, 1000010, 1000011,
            ...
            1006038, 1006039, 1006040, 1007021, 1007028, 1007081, 1007091,
            1007092, 1007095, 1007099],
           dtype='int64', name='User_ID', length=5885)

In [143]:
user_pivot.head()

Product_Category  1  2  3   4  5  6   7  8  9  10  11  12  13  14
User_ID                                                          
1000001           0  0  3   2  0  0   7  0  0   0   0   0   0   0
1000002           4  0  0  12  1  0  20  0  0   0   0   0   0   0
1000003           1  0  0   8  0  0   1  0  0   0   0   0   0   0
1000005           5  0  0  16  2  1  29  0  0   1   0   0   1   0
1000006           0  1  0  11  1  0   6  0  0   0   0   0   0   0

In [144]:
user_pivot.index.names

FrozenList(['User_ID'])

In [145]:
# To make manipulations on data, we shall convert columns of categorical index to string type
user_pivot.columns = user_pivot.columns.astype('str')
user_pivot = user_pivot.reset_index()
user_pivot.head()

Product_Category  User_ID  1  2  3   4  5  6   7  8  9  10  11  12  13  14
0                 1000001  0  0  3   2  0  0   7  0  0   0   0   0   0   0
1                 1000002  4  0  0  12  1  0  20  0  0   0   0   0   0   0
2                 1000003  1  0  0   8  0  0   1  0  0   0   0   0   0   0
3                 1000005  5  0  0  16  2  1  29  0  0   1   0   0   1   0
4                 1000006  0  1  0  11  1  0   6  0  0   0   0   0   0   0

In [146]:
user_pivot.columns

Index(['User_ID', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11',
       '12', '13', '14'],
      dtype='object', name='Product_Category')

#### Pivot - Wide to Long format

In [147]:
melted = pd.melt(user_pivot, ['User_ID'])
melted.head(10)

User_ID Product_Category  value
0  1000001                1      0
1  1000002                1      4
2  1000003                1      1
3  1000005                1      5
4  1000006                1      0
5  1000007                1      0
6  1000008                1      3
7  1000009                1      2
8  1000010                1      2
9  1000011                1      0

In [148]:
melted[melted.value != 0].sort_values(['User_ID', 'Product_Category']).head(10)

User_ID Product_Category  value
11770  1000001                3      3
17655  1000001                4      2
35310  1000001                7      7
1      1000002                1      4
17656  1000002                4     12
23541  1000002                5      1
35311  1000002                7     20
2      1000003                1      1
17657  1000003                4      8
35312  1000003                7      1

### Custom functions

In [149]:
user_pivot = user_pivot.set_index("User_ID")
user_pivot.head(5)

Product_Category  1  2  3   4  5  6   7  8  9  10  11  12  13  14
User_ID                                                          
1000001           0  0  3   2  0  0   7  0  0   0   0   0   0   0
1000002           4  0  0  12  1  0  20  0  0   0   0   0   0   0
1000003           1  0  0   8  0  0   1  0  0   0   0   0   0   0
1000005           5  0  0  16  2  1  29  0  0   1   0   0   1   0
1000006           0  1  0  11  1  0   6  0  0   0   0   0   0   0

In [150]:
# Writing a custom function to identify if user has a favourite product category
def favorite(lst):
    f=np.max(lst)/np.sum(lst)
    return(f)

user_pivot['val'] = pd.DataFrame(user_pivot.apply(favorite, axis=1))

In [151]:
user_pivot.loc[user_pivot.val >= 0.6, 'Fav'] = 1 
user_pivot.loc[user_pivot.val <  0.6, 'Fav'] = 0 
user_pivot.head(10)

Product_Category  1  2  3   4  5  6   7  8  9  10  11  12  13  14       val  \
User_ID                                                                       
1000001           0  0  3   2  0  0   7  0  0   0   0   0   0   0  0.583333   
1000002           4  0  0  12  1  0  20  0  0   0   0   0   0   0  0.540541   
1000003           1  0  0   8  0  0   1  0  0   0   0   0   0   0  0.800000   
1000005           5  0  0  16  2  1  29  0  0   1   0   0   1   0  0.527273   
1000006           0  1  0  11  1  0   6  0  0   0   0   0   0   0  0.578947   
1000007           0  0  0   1  0  0   0  0  0   1   0   0   0   0  0.500000   
1000008           3  0  0   7  0  0  39  0  0   0   0   0   0   0  0.795918   
1000009           2  0  0   7  0  0  20  0  0   0   0   0   0   0  0.689655   
1000010           2  4  2  79  0  0  33  0  0   0   0   1   0   1  0.647541   
1000011           0  1  0  28  0  0  11  0  0   0   0   0   0   0  0.700000   

Product_Category  Fav  
User_ID                
1000001           0.0  
1000002           0.0  
1000003           1.0  
1000005           0.0  
1000006           0.0  
1000007           0.0  
1000008           1.0  
1000009           1.0  
1000010           1.0  
1000011           1.0

## Learning Outcomes

* Be able to process, analyse and manipulate using Pandas .
    * Be able to create pandas series and Dataframe
    * Be able to subset data frames using pandas functions
    * Be able to use pandas functions and methods on data frames to understand data and describe it
* Be able to implement exploratory data analysis by applying required data preprocessing functions
    * Be able to read data from different file formats
    * Be able to differentiate numerical and categorical features and explicit type conversions if necessary
    * Be able to use and apply functions for duplicates, missing value and extreme value analysis
    * Be able to implement explicit data manipulation like standardization discretization and dummification which 
      might be necessary for certain ML algorithms
    * Be able to implement data aggregations and data merging
    * Be able to handle date manipulations for data analysis